exp/zapslog/slog.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2023 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:build zapslog
+
+package zapslog
+
+import (
+	"context"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"golang.org/x/exp/slog"
+)
+
+type handler struct {
+	logger *zap.Logger
```

Comment:
@knight42 I think there was a misunderstanding.
We want to implement this interface around zapcore.Core.

slog.Handler and zapcore.Core are at the same level
similar to how slog.Logger and zap.Logger are at the same level.

Reply:
@abhinav Oh thanks for pointing it out, so what we actually provide in this package is sth like [slog.JSONHandler](https://pkg.go.dev/golang.org/x/exp/slog#JSONHandler), right? Should we provide a similar [HandlerOptions](https://pkg.go.dev/golang.org/x/exp/slog#HandlerOptions) struct? 

What I currently have in mind is defining a public function in the package like the following, does it need further improvement?
```go
func NewHandlerFromCore(core zapcore.Core) slog.Handler {
  ...
}
```

Reply:
Yeah, that sounds right, @knight42.
Suggested changes:

- We don't need "FromCore" there since it'll be defined in zapslog, so that we're dealing with Zap abstractions will already be in the name's context.
- We could maybe return a `*Handler` struct instead of an interface should we need flexibility there in the future.

So perhaps:

```go
package zapslog

type Handler struct{ ... }

var _ slog.Handler = (*Handler)(nil)

func NewHandler(core zapcore.Core) *Handler
```

(In case you're not aware, that `var _` line is a [compile-time check to verify interface compliance](https://github.com/uber-go/guide/blob/master/style.md#verify-interface-compliance).)